src/markdown/blocks/heading.js

```diff
@@ -91,9 +91,39 @@ function parseHeadingText(state, level, initialText) {
         data = { id };
     }
 
+    const trimmedNodes = newState.nodes.map((node, i) => {
+        let newNode = node;
+        const leaves = node.getLeaves();
```

Comment:
Don't list leaves if the node is not the first or the last.

And why iterating all blocks if you just want to update the first and last one ?